core/src/toga/documents.py

```diff
@@ -118,5 +153,332 @@ def create(self) -> None:
+
+        self.elements: list[Document] = []
+
+    @property
+    def types(self) -> list[type[Document]]:
+        """The list of document types the app can manage.
+
+        The first document type in the list is the app's default document type.
+        """
+        return self._types
+
+    def __iter__(self) -> Iterator[Document]:
+        return iter(self.elements)
+
+    def __contains__(self, value: object) -> bool:
+        return value in self.elements
+
+    def __len__(self) -> int:
+        return len(self.elements)
+
+    def __getitem__(self, path_or_index):
+        # Look up by index
+        if isinstance(path_or_index, int):
+            return self.elements[path_or_index]
+
+        # Look up by path
+        if sys.version_info < (3, 9):  # pragma: no-cover-if-gte-py39
+            # resolve() *should* turn the path into an absolute path;
+            # but on Windows, with Python 3.8, it doesn't.
+            path = Path(path_or_index).absolute().resolve()
+        else:  # pragma: no-cover-if-lt-py39
+            path = Path(path_or_index).resolve()
+        for item in self.elements:
+            if item.path == path:
+                return item
+
+        # No match found
+        raise KeyError(path_or_index)
+
+    def _add(self, document: Path):
+        if document in self:
+            raise ValueError("Document is already being managed.")
+
+        self.elements.append(document)
+
+    def _remove(self, document: Path):
+        if document not in self:
+            raise ValueError("Document is not being managed.")
+
+        self.elements.remove(document)
+
+    def new(self, document_type: type[Document]) -> Document:
+        """Create a new document of the given type, and show the document window.
+
+        :param document_type: The document type that has been requested.
+        :returns: The newly created document.
+        """
+        document = document_type(app=self.app)
+        document.show()
+        return document
+
+    async def request_open(self) -> Document:
+        """Present a dialog asking the user for a document to open, and pass the
+        selected path to :meth:`open`.
+
+        :returns: The document that was opened.
+        :raises ValueError: If the path describes a file that is of a type that doesn't
+            match a registered document type.
+        """
+        # A safety catch: if app modal dialogs aren't actually modal (eg, macOS) prevent
+        # a second open dialog from being opened when one is already active. Attach the
+        # dialog instance as a private attribute; delete as soon as the future is
+        # complete.
+        if hasattr(self, "_open_dialog"):
+            return
+
+        # CLOSE_ON_LAST_WINDOW is a proxy for the GTK/Windows behavior of loading content
+        # into the existing window. This is actually implemented by creating a new window
+        # and disposing of the old one; mark the current window for cleanup
```

Comment:
I agree that a document app should definitely use (or, at least, be able to use) multiple windows, regardless of platform. 

There's an edge case you're missing in your analysis, though - New. On GTK/Windows, File > New always creates a new window; File > Open replaces the existing window. 

From my survey of platform behavior when I set this up, on Windows, most apps that I saw (e.g., notepad) have distinct "New" and "New Window" commands, with "New" being a replacing operation by default. That's a little difficult to accomodate cleanly when there are potentially multiple document types, because you end up with 2 New commands for every document type, so your File menu gets busy really quickly. Open is always a replacing operation.

The behaviour is less consistent on GTK (shocking nobody), but most apps (even Terminal) have an app-level "New window" command in addition to an "open" that has replacing behavior; if there's a "new document" option, it often creates a second *tab* in the same window (which isn't something we're in a place to support anyway).

So - I opted for a single "New" that is effectively "new window" on all platforms. For the "single document window" case, this seems the most consistent without adding an explicit second new command; and in a multiple-document window (such as an IDE), it should be possible to incorporate "open tab rather than replace" behavior with some relatively minor affordances in the Document class for the multi-document app, plus a user-added "New window" command. 

It's also *possible* for a "single document window" app on GTK/Windows to circumvent the replacement behavior by deleting the `_replace` attribute of the current window before creating the new window in the Document's `create()` method. This isn't an official API - we'd possibly want to find a cleaner way to formalise this if we want to make it official - but it's at least *possible*.

The other possible missing piece that GTK and Windows don't currently have is the analog of "drag document onto app" or "double click on icon". On macOS, this opens a new document window in the same app instance; I'm not sure what (if any) mechanisms exist to make this sort of behavior work on GTK and Windows, but in the meantime, you'll get 2 app instances with distinct document windows.

Regarding the larger "meta" point - I'm not sure I agree that cross-platform consistency is as important as platform-native consistency, provided the consequences for the developer aren't invasive. Replace behavior feels alien on macOS; non-replace behavior feels alien on macOS and GTK. Given that platform consistency is one of the key design principles of Toga, and we've got a reasonably clean way of accomodating both in a way that only affects how the user *uses* the app, not how it is built, it seems reasonable to me to retain the platform consistency.

And - as a note related to your other review comment about merging `create` and `open` - it occurs to me that there's possibly an option here for formalising how `replace` works by making `replace` an optional method on the Document class that is only invoked if it is defined and the platform supports replacing behaviour, and then pass the replacement request to the creation of the document. I don't think we need to resolve that design decision right now - I think it fits better in a discussion about formally adding multi-document apps - but it's worth noting that this approach wouldn't be possible if `create` and `open` were merged.